app/models/manageiq/providers/redfish/physical_infra_manager/operations/power.rb

```diff
@@ -40,26 +40,62 @@ def restart_mgmt_controller(_server, _options)
       $redfish_log.error("Restarting management controller is not supported.")
     end
 
+    # Select any supported method of powering the server down.
+    def power_down(server, prefer_force = false)
+      rtypes = %w[GracefulShutdown ForceOff]
+      rtypes.reverse! if prefer_force
+      execute_first_valid_reset_action(server, rtypes)
+    end
+
+    # Select any supported method of powering the server up.
+    def power_up(server, prefer_force = false)
+      rtypes = %w[On ForceOn]
+      rtypes.reverse! if prefer_force
+      execute_first_valid_reset_action(server, rtypes)
+    end
+
     private
 
-    def reset_server(server, reset_type)
-      $redfish_log.info("Requesting #{reset_type} for #{server.ems_ref}.")
-      with_provider_connection do |client|
-        system = client.find(server.ems_ref)
-        if system.nil?
-          $redfish_log.error("#{server.ems_ref} does not exist anymore.")
-          return
+    def execute_first_valid_reset_action(server, rtypes)
+      with_system(server) do |system|
+        available_rtypes = get_available_rtypes(system)
+        rtype = rtypes.select { |t| available_rtypes.include?(t) }.first
+        if rtype.nil?
+          $redfish_log.error("#{rtypes} and ${available_rtypes} are disjunct.")
+          # TOOD(@tadeboro): Raise an exception here?
+        else
+          execute_reset_action(system, rtype)
         end
+      end
+    end
 
-        response = system.Actions["#ComputerSystem.Reset"].post(
-          :field => "target", :payload => { "ResetType" => reset_type }
-        )
-        if [200, 202, 204].include?(response.status)
-          $redfish_log.info("#{reset_type} for #{server.ems_ref} started.")
+    def with_system(server)
+      server.with_provider_object do |system|
+        if system.nil?
+          $redfish_log.error("#{server.ems_ref} does not exist.")
         else
-          $redfish_log.error("#{reset_type} for #{server.ems_ref} failed.")
+          yield system
         end
       end
     end
+
+    def get_available_rtypes(system)
+      reset = system.Actions["#ComputerSystem.Reset"]
+      reset["ResetType@Redfish.AllowableValues"] ||
+        reset["@Redfish.ActionInfo"].Parameters.AllowableValues ||
```

Comment:
Isn't this an array?

Reply:
Yes, it is an array that needs to be filtered for a proper parameter, but I pushed the broken code because I wanted some feedback about the general direction (which I got, thanks).